### PR TITLE
go and pypy

### PR DESCRIPTION
**Pull request by boromil**
I've added a Go implementation and PyPy run.sh line.
You're welcome to use it :)

### CODE REVIEW

File: primes.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+func getPrimes7(n int) []int {
+	if n < 2 {
+		return []int{}
+	}
+	if n == 2 {
+		return []int{2}
+	}
+
+	s := make([]int, 0, n/2)
```
Comment (famzah) [inline primes.go]: You hint the compiler that we will use "n/2" elements in "s".

File: primes.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			for j < half {
+				s[j] = 0
+				j += m
+			}
+		}
+		i++
+		m = 2*i + 3
+	}
+
+	res := make([]int, 0, n/int(math.Log(float64(n))))
```
Comment (famzah) [inline primes.go]: You hint the compiler about the size of the array, here too.

**Comment by famzah** [comment]
Hi, thanks for the contribution.

I've commented inline in the pull request two lines of code. They hint the compiler about the size of the array which I find "unfair" :) Since it's the dynamic array allocation & performance that we want to benchmark, I'd rather not hint the compiler about anything. If the algorithm we were about to implement wasn't the "Prime number theorem", we may not even had the chance to know the size prematurely (for example if we were expecting an unknown count of user input).

Long story short - is there a way to declare an array with an unknown size in Go? Something like the following:

``` go
s = []int{}
```

**Comment by boromil** [comment]
> Long story short - is there a way to declare an array with an unknown size in Go? Something like the following:
> 
> ``` go
> s = []int{}
> ```

you almost had it :)

``` go
s := []int{}
```

will work

I've made the necessary changes on my fork.

**Comment by famzah** [comment]
Thanks a lot for your response. In a couple of days I'll merge your code and re-do the tests + publish the results.

**Comment by famzah** [comment]
I won't be able to update the website results with the Go implementation. The setup I did the tests is no longer available, and due to personal tasks I can't spend more time to recreate it, in order to have all the languages and versions installed again.

From what I can say, Go performs as fast as Nodejs:

```
== JavaScript (nodejs) ==                                                                           

real    0m4.365s                                                                                    
user    0m4.180s                                                                                    
sys     0m0.236s                                                                                    

real    0m4.301s                                                                                    
user    0m4.132s                                                                                    
sys     0m0.260s                                                                                    

v0.10.25                                                                                            

== Go (not optimized, default compiler) ==                                                          

real    0m3.819s                                                                                    
user    0m4.192s                                                                                    
sys     0m0.128s                                                                                    

real    0m3.222s                                                                                    
user    0m4.332s                                                                                    
sys     0m0.100s                                                                                    
```

I'm a little surprised to see that Go is using some multi-threaded approach. The "real" wall-clock time is less than the sum of User + System CPU times.

And PyPy is almost as fast as C++:

```
== C++ (optimized with -O2) ==                                                                      

real    0m1.241s                                                                                    
user    0m1.104s                                                                                    
sys     0m0.128s                                                                                    

real    0m1.244s                                                                                    
user    0m1.092s                                                                                    
sys     0m0.148s                                                                                    

g++ (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4                                                           

== PyPy 2.7 ==                                                                                      

real    0m1.642s                                                                                    
user    0m1.484s                                                                                    
sys     0m0.152s                                                                                    

real    0m1.637s                                                                                    
user    0m1.456s                                                                                    
sys     0m0.176s                                                                                    

Python 2.7.3 (2.2.1+dfsg-1ubuntu0.3, Sep 30 2015, 15:18:27)                                         
[PyPy 2.2.1 with GCC 4.8.4]                                                                         
```

The PyPy results are consistent with the comments at the blog page.

**Comment by boromil** [comment]
The default Go compiler isn't relay build for run-time performance, you might get better results with [gccgo](https://golang.org/doc/install/gccgo) and _-O2_ compile option - similar to C++.

PyPy is awesome (and does some crazy black magic under the hood) :)

Would be interesting to see some memory usage when running these benchmarks.
